app/src/server/utils/stripe.ts

```diff
@@ -0,0 +1,144 @@
+import Stripe from "stripe";
+import { env } from "~/env.mjs";
+import { prisma } from "~/server/db";
+import { TRPCError } from "@trpc/server";
+import type { Decimal } from "@prisma/client/runtime";
+
+const stripe = new Stripe(env.STRIPE_SECRET_KEY ?? "");
+
+export async function createStripePaymentIntent({
+  amount,
+  stripeCustomerId,
+  paymentMethodId,
+  returnUrl,
+  invoiceId,
+}: {
+  amount: number;
+  stripeCustomerId: string;
+  paymentMethodId: string;
+  returnUrl: string;
+  invoiceId: string;
+}) {
+  return stripe.paymentIntents.create({
+    amount,
+    currency: "usd",
+    automatic_payment_methods: { enabled: true },
+    customer: stripeCustomerId,
+    payment_method: paymentMethodId,
+    return_url: returnUrl,
+    off_session: true,
+    confirm: true,
+    metadata: {
+      invoiceId, // Used to track status using webhooks
+    },
+  });
+}
+export async function createSetupIntent(stripeCustomerId: string) {
+  return stripe.setupIntents.create({
+    customer: stripeCustomerId,
+    usage: "off_session",
+    automatic_payment_methods: {
+      enabled: true,
+    },
+  });
+}
+
+export async function getStripeCustomer(stripeCustomerId: string) {
+  return stripe.customers.retrieve(stripeCustomerId);
+}
+export async function deletePaymentMethod(paymentMethodId: string) {
+  return stripe.paymentMethods.detach(paymentMethodId);
+}
+
+export async function getDefaultPaymentMethodId(stripeCustomerId: string) {
+  const customer = await stripe.customers.retrieve(stripeCustomerId);
+
+  if (customer.deleted) {
+    return undefined;
+  }
+
+  const paymentMethodId = customer.invoice_settings.default_payment_method;
+  if (!paymentMethodId || typeof paymentMethodId !== "string") {
+    return undefined;
+  }
+
+  return paymentMethodId;
+}
+
+export async function setDefaultPaymentMethod(stripeCustomerId: string, paymentMethodId: string) {
+  return await stripe.customers.update(stripeCustomerId, {
+    invoice_settings: {
+      default_payment_method: paymentMethodId,
+    },
+  });
+}
+
+export async function getPaymentMethods(stripeCustomerId: string) {
+  const customer = await stripe.customers.retrieve(stripeCustomerId);
+
+  if (customer.deleted) {
+    return undefined;
+  }
+
+  return await stripe.paymentMethods.list({
+    customer: stripeCustomerId,
+    type: "card",
+  });
+}
+
+export function usdToCents(usd: number | Decimal) {
+  const cents = Math.round(Number(usd) * 100);
+  return cents >= 0 ? cents : 0;
+}
+
+export async function createStripeCustomerAndConnectItToProject(projectId: string) {
+  // Find a project
+  const project = await prisma.project.findUnique({
+    where: {
+      id: projectId,
+    },
+  });
+
```

Comment:
I would check to ensure that project exists here and throw a 404 if it doesn't. That way you don't need all the `project?.` later in the function, and we don't create customers for nonexistent projects.

Reply:
Completely agree; changed it.